crates/block-producer/src/block_builder/mod.rs

```diff
@@ -72,37 +73,161 @@ impl BlockBuilder {
+        let (block_number, batches) = mempool.lock().await.select_block();
+        SelectedBlock { block_number, batches }
+    }
 
-            tokio::time::sleep(proving_duration).await;
+    #[instrument(target = COMPONENT, name = "block_builder.get_block_inputs", skip_all, err)]
+    async fn get_block_inputs(
+        &self,
+        selected_block: SelectedBlock,
+    ) -> Result<BlockSummaryAndInputs, BuildBlockError> {
+        let SelectedBlock { block_number: _, batches } = selected_block;
+        let summary = BlockSummary::summarize_batches(&batches);
 
-            // Randomly inject failures at the given rate.
-            //
-            // Note: Rng::gen rolls between [0, 1.0) for f32, so this works as expected.
-            if rand::thread_rng().gen::<f32>() < self.failure_rate {
-                result = Err(BuildBlockError::InjectedFailure);
-            }
+        let inputs = self
+            .store
+            .get_block_inputs(
+                summary.updated_accounts.iter().copied(),
+                summary.nullifiers.iter(),
+                summary.dangling_notes.iter(),
+            )
+            .await
+            .map_err(BuildBlockError::GetBlockInputsFailed)?;
 
-            let mut mempool = mempool.lock().await;
-            match result {
-                Ok(_) => mempool.block_committed(block_number),
-                Err(_) => mempool.block_failed(block_number),
-            }
+        let missing_notes: Vec<_> = summary
+            .dangling_notes
+            .difference(&inputs.found_unauthenticated_notes.note_ids())
+            .copied()
+            .collect();
+        if !missing_notes.is_empty() {
+            return Err(BuildBlockError::UnauthenticatedNotesNotFound(missing_notes));
         }
+
+        Ok(BlockSummaryAndInputs { batches, summary, inputs })
     }
 
-    #[instrument(target = COMPONENT, skip_all, err)]
-    async fn build_block(&self, batches: &[ProvenBatch]) -> Result<(), BuildBlockError> {
-        info!(
-            target: COMPONENT,
-            num_batches = batches.len(),
-            batches = %format_array(batches.iter().map(ProvenBatch::id)),
-        );
+    #[instrument(target = COMPONENT, name = "block_builder.prove_block", skip_all, err)]
+    async fn prove_block(
+        &self,
+        preimage: BlockSummaryAndInputs,
+    ) -> Result<ProvenBlock, BuildBlockError> {
+        let BlockSummaryAndInputs { batches, summary, inputs } = preimage;
+
+        let (block_header_witness, updated_accounts) = BlockWitness::new(inputs, &batches)?;
 
-        let updated_account_set: BTreeSet<AccountId> = batches
+        let new_block_header = self.block_kernel.prove(block_header_witness)?;
+
+        let block = Block::new(
+            new_block_header,
+            updated_accounts,
+            summary.output_notes,
+            summary.nullifiers,
+        )
+        .expect("invalid block components");
```

Comment:
Should we describe why this cannot fail or is this temporary?

Reply:
I've turned this into a new error variant instead.